File: Data Structures & Algorithms/reorder-linked-list/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        s, f = head, head.next
        while f and f.next:
            s = s.next
            f = f.next.next
        if not s.next: return
        prev, curr, nxt = s.next, s.next, s.next.next
        s.next, curr.next = None, None

        while nxt:
            curr = nxt
            nxt = nxt.next
            curr.next = prev
            prev = curr

        l1, l2 = head, curr
        while l1 and l2:
            nxt = l1.next
            l1.next = l2
            l1 = nxt
            nxt = l2.next
            l2.next = l1
            l2 = nxt

File: Data Structures & Algorithms/reverse-linked-list-ii/submission-0.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseBetween(ListNode* head, int left, int right) {
        ListNode *prev, *curr, *nxt, *begin, *before_begin;
        prev = nullptr;
        nxt = curr = head;
        int left_cpy = left;
        
        while (left_cpy > 1) {
            nxt = nxt->next;
            prev = curr;
            curr = nxt;
            left_cpy--;
        }
        begin = curr;
        before_begin = prev;

        int reverseTimes = right - left + 1;
        while (curr && reverseTimes > 0) {
            nxt = nxt->next;
            curr->next = prev;
            prev = curr;
            curr = nxt;
            reverseTimes--;
        }
        begin->next = curr;
        if (before_begin) before_begin->next = prev;
        if (left == 1) head = prev;

        return head;
    }
};